File: apps/frontend/src/api/schemas.ts
import { z } from 'zod';

// Meta information for API responses
export const MetaSchema = z.object({
  total: z.number().optional(),
  page: z.number().optional(),
  pageSize: z.number().optional(),
  totalPages: z.number().optional(),
});

// Generic paginated response
export const PagedSchema = <T extends z.ZodTypeAny>(itemSchema: T) => z.object({
  data: z.array(itemSchema),
  meta: MetaSchema,
});

// Site/Facility schemas
export const SiteSchema = z.object({
  site_id: z.number(),
  site_code: z.string(),
  site_name: z.string(),
});

// Program schema
export const ProgramSchema = z.object({
  program_id: z.number(),
  program_name: z.string(),
});

// Subprogram schema  
export const SubprogramSchema = z.object({
  program_id: z.number(),
  subprogram_id: z.number(),
  subprogram_name: z.string(),
});

// Facility Summary schema
export const FacilitySummarySchema = z.object({
  total_sites: z.number(),
  total_programs: z.number(),
  total_beds: z.number(),
  avg_occupancy: z.number(),
  kpis: z.record(z.string(), z.any()).optional(),
});

// Patient schema
export const PatientSchema = z.object({
  patient_id: z.number(),
  age: z.number(),
  gender: z.enum(['Male', 'Female', 'Other']),
  home_lha: z.string(),
  facility_site_code: z.string(),
  admission_date: z.string().optional(),
});

// ED Encounter schema
export const EdEncounterSchema = z.object({
  encounter_id: z.number(),
  patient_id: z.number(),
  arrival_datetime: z.string(),
  ed_subservice: z.enum(['Adult ED', 'Pediatric ED', 'Urgent Care Centre']),
  acuity: z.enum(['CTAS_1', 'CTAS_2', 'CTAS_3', 'CTAS_4', 'CTAS_5']),
  disposition: z.enum(['Discharged', 'Admitted', 'Transferred', 'LWBS', 'Expired']),
});

// Projection Point schema
export const ProjectionPointSchema = z.object({
  year: z.number(),
  month: z.number().optional(),
  value: z.number(),
  metric: z.string(),
  site_code: z.string().optional(),
});

// Scenario-related schemas for the 3 screens
export const ScenarioParamsSchema = z.object({
  occupancy_target: z.number().min(0.8).max(1.0),
  los_delta: z.number().min(-0.5).max(0.5),
  alc_target: z.number().min(0.0).max(0.5),
  growth_pct: z.number().min(-0.2).max(0.2),
  schedule_code: z.string().default('Sched-A'),
  seasonality: z.boolean().default(false),
});

export const ScenarioRequestSchema = z.object({
  sites: z.array(z.number()).min(1),
  program_id: z.number(),
  baseline_year: z.number().default(2022),
  horizon_years: z.number().default(3),
  params: ScenarioParamsSchema,
});

export const SiteResultSchema = z.object({
  site_id: z.number(),
  site_code: z.string(),
  site_name: z.string(),
  admissions_projected: z.number(),
  los_effective: z.number(),
  patient_days: z.number(),
  census_average: z.number(),
  required_beds: z.number(),
  staffed_beds: z.number(),
  capacity_gap: z.number(),
  nursing_fte: z.number().nullable().optional(),
});

export const ScenarioKPIsSchema = z.object({
  total_required_beds: z.number(),
  total_staffed_beds: z.number(),
  total_capacity_gap: z.number(),
  total_nursing_fte: z.number().nullable().optional(),
  avg_occupancy: z.number(),
  total_admissions: z.number(),
  avg_los_effective: z.number(),
});

export const ScenarioResponseSchema = z.object({
  kpis: ScenarioKPIsSchema,
  by_site: z.array(SiteResultSchema),
  metadata: z.record(z.string(), z.any()).default({}),
});

// API Response wrappers
export const ApiResponseSchema = <T extends z.ZodTypeAny>(dataSchema: T) => z.object({
  data: dataSchema,
  meta: z.record(z.string(), z.any()).default({}),
});

File: api/schemas.py
"""
Pydantic schemas for the extended API.
"""

from datetime import datetime
from typing import List, Optional, Dict, Any, Union
from pydantic import model_validator
from pydantic import BaseModel, Field
from decimal import Decimal


# Base response models
class ApiResponse(BaseModel):
    """Standard API response format."""
    data: Any
    meta: Dict[str, Any] = Field(default_factory=dict)


class PaginatedResponse(BaseModel):
    """Paginated API response format.""" 
    data: List[Any]
    meta: Dict[str, Any] = Field(default_factory=dict)


# Reference table models
class StaffedBedsSchedule(BaseModel):
    """Staffed beds schedule model."""
    id: int
    site_id: int
    program_id: int
    schedule_code: str
    staffed_beds: int


class ClinicalBaseline(BaseModel):
    """Clinical baseline model."""
    id: int
    site_id: int
    program_id: int
    baseline_year: int
    los_base_days: float
    alc_rate: float


class SeasonalityMonthly(BaseModel):
    """Seasonality monthly model."""
    id: int
    site_id: Optional[int]
    program_id: Optional[int]
    month: int
    multiplier: float


class StaffingFactor(BaseModel):
    """Staffing factor model."""
    id: int
    program_id: int
    subprogram_id: Optional[int]
    hppd: float
    annual_hours_per_fte: int
    productivity_factor: float


# Dimension models (for reference endpoints)
class Site(BaseModel):
    """Site dimension model."""
    site_id: int
    site_code: str
    site_name: str


class Program(BaseModel):
    """Program dimension model."""
    program_id: int
    program_name: str


class Subprogram(BaseModel):
    """Subprogram dimension model."""
    program_id: int
    subprogram_id: int
    subprogram_name: str


# Scenario calculation models
class ScenarioParams(BaseModel):
    """Parameters for scenario calculations."""
    occupancy_target: float = Field(..., ge=0.80, le=1.0, description="Target occupancy rate")
    los_delta: float = Field(0.0, ge=-0.50, le=0.50, description="LOS change percentage")
    alc_target: float = Field(..., ge=0.0, le=0.50, description="Target ALC rate")
    growth_pct: float = Field(0.0, ge=-0.20, le=0.20, description="Annual growth percentage")
    schedule_code: str = Field("Sched-A", description="Capacity schedule code")
    seasonality: bool = Field(False, description="Apply seasonality adjustments")


class ScenarioRequest(BaseModel):
    """Request for scenario calculation."""
    sites: List[int] = Field(..., description="Site IDs to include")
    # Backwards-compatible: either provide a single `program_id` or a
    # list `program_ids`. Internally we normalize to `program_ids`.
    program_id: Optional[int] = Field(None, description="Program ID (deprecated, use program_ids)")
    program_ids: Optional[List[int]] = Field(None, description="List of Program IDs to include in the calculation")
    baseline_year: int = Field(2022, description="Baseline year")
    horizon_years: int = Field(3, description="Planning horizon in years")
    params: ScenarioParams

    @model_validator(mode='before')
    def ensure_program_ids(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        # If caller provided program_id but not program_ids, normalize to program_ids
        pid = values.get('program_id')
        pids = values.get('program_ids')
        if not pids and pid is None:
            raise ValueError('Either program_id or program_ids must be provided')
        if not pids and pid is not None:
            values['program_ids'] = [pid]
        # If program_ids provided but program_id missing, keep program_id as the first value
        if pids and not pid:
            try:
                values['program_id'] = pids[0]
            except Exception:
                pass
        return values


class SiteResult(BaseModel):
    """Results for a single site."""
    site_id: int
    site_code: str
    site_name: str
    admissions_projected: int
    los_effective: float
    patient_days: int
    census_average: float
    required_beds: int
    staffed_beds: int
    capacity_gap: int
    nursing_fte: Optional[float] = None


class ScenarioKPIs(BaseModel):
    """Aggregate scenario KPIs."""
    total_required_beds: int
    total_staffed_beds: int
    total_capacity_gap: int
    total_nursing_fte: Optional[float] = None
    avg_occupancy: float
    total_admissions: int
    avg_los_effective: float


class ScenarioResponse(BaseModel):
    """Response for scenario calculation."""
    kpis: ScenarioKPIs
    by_site: List[SiteResult]
    metadata: Dict[str, Any] = Field(default_factory=dict)


# Error models
class ErrorDetail(BaseModel):
    """Error detail model."""
    type: str
    message: str
    field: Optional[str] = None


class ErrorResponse(BaseModel):
    """Error response model.""" 
    error: str
    details: List[ErrorDetail] = Field(default_factory=list)
    request_id: Optional[str] = None